batch_writer.go

```diff
@@ -47,7 +48,76 @@ func (w *BatchWriter) Initialize() {
 	w.logger = logrus.WithField("tag", "batch_writer")
 }
 
+func (w *BatchWriter) WriteRowBatchWithConcurrency(concurrency int) func(batch *RowBatch) error {
+	w.logger.WithField("concurrency", concurrency).Infof("running batch writer with concurrency")
+
+	return func(batch *RowBatch) error {
+		wg := &sync.WaitGroup{}
+
+		splitBatches, err := batch.Split(concurrency)
+		if err != nil {
+			return err
+		}
+
+		wg.Add(len(splitBatches))
+
+		for _, splitBatch := range splitBatches {
+			go func(batch *RowBatch) error {
+				defer wg.Done()
+
+				err := w.writeRowBatch(batch)
+				if err != nil {
+					panic(err) // TODO: Handle this error
+				}
+
+				return nil
+			}(splitBatch)
+		}
+
+		wg.Wait()
+
+		err = w.saveBatchProgressToState(batch)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
 func (w *BatchWriter) WriteRowBatch(batch *RowBatch) error {
+	w.logger.Infof("running batch writer with no concurrency")
+
+	err := w.writeRowBatch(batch)
+	if err != nil {
+		return err
+	}
+
+	err = w.saveBatchProgressToState(batch)
```

Comment:
Note for myself next review, the previous code has saving the progress in the `WithRetries` function. This effectively moves it out. I _think_ this is okay, but should probably verify in the next round of review (as I'm more focused on the `Split` function right now).